fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -1,155 +1,175 @@
-                if let Some(signed_block) = db::load_block(&db, index).await {
-                    outgoing_message_sender
-                        .send((Message::Block(signed_block), Recipient::Peer(peer_id)))
-                        .await
-                        .ok();
+    network_data_receiver: async_channel::Receiver<Vec<u8>>,
+    signed_block_receiver: async_channel::Receiver<SignedBlock>,
+    relay_handle: JoinHandle<()>,
+}
 
-                    tracing::info!("Served the block with index {} to peer {}", index, peer_id);
+impl AtomicBroadcast {
+    /// This function starts the atomic broadcast instance. A running instance
+    /// serves signed blocks to peers on request even if we do not run a
+    /// session.
+    pub fn new(
+        keychain: Keychain,
+        db: Database,
+        mempool_item_receiver: async_channel::Receiver<Vec<u8>>,
+        incoming_message_receiver: async_channel::Receiver<(Message, PeerId)>,
+        outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+    ) -> Self {
+        let (network_data_sender, network_data_receiver) = async_channel::bounded(256);
+        let (signed_block_sender, signed_block_receiver) = async_channel::bounded(16);
+
+        let db_clone = db.clone();
+        let sender_clone = outgoing_message_sender.clone();
+
+        let relay_handle = tokio::spawn(async move {
+            while let Ok((message, peer_id)) = incoming_message_receiver.recv().await {
+                match message {
+                    Message::NetworkData(network_data) => {
+                        // if we were to await a send the relay loop may get stuck if we are not
+                        // running a session
+                        network_data_sender.try_send(network_data).ok();
+                    }
+                    Message::BlockRequest(index) => {
+                        if let Some(signed_block) = db::load_block(&db_clone, index).await {
+                            sender_clone
+                                .send((Message::Block(signed_block), Recipient::Peer(peer_id)))
+                                .await
+                                .ok();
+
+                            tracing::info!(
+                                "Served the block with index {} to peer {}",
+                                index,
+                                peer_id
+                            );
+                        }
+                    }
+                    Message::Block(signed_block) => {
+                        // if we were to await a send the relay loop may get stuck if we are not
+                        // running a session
+                        signed_block_sender.try_send(signed_block).ok();
+                    }
                 }
             }
-            Message::Block(signed_block) => {
-                signed_block_sender.send(signed_block).await.ok();
-            }
+
+            std::future::pending().await
```

Comment:
nit: why not return instead of blocking forever? We might also want to log that we exited the relay loop.

Reply:
If we would return here the current aleph session would shut down as the senders get dropped. We only want to shut it down though if the ordered_item_receiver is dropped.